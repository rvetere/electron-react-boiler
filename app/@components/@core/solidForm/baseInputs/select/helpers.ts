import { ISelectOption } from './types'

export const getSelectedOption = (givenValue: string | number | null, options: ISelectOption[]): ISelectOption | undefined =>
  options.find(({ value }: ISelectOption): boolean => value === givenValue)
