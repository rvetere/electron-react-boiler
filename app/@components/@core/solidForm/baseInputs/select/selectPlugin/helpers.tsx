import { ChevronDown, ChevronUp, Close } from '@core/image/icon'
import { ISelectOption } from '../types'

export const ArrowRenderer = ({ isOpen }: { isOpen: boolean }): JSX.Element => {
  if (isOpen) {
    return <ChevronUp />
  }

  return <ChevronDown />
}

export const ClearRenderer = (): JSX.Element => <Close />

export const unClearableOptions = (options?: ISelectOption[]): ISelectOption[] => {
  if (!options) {
    return []
  }

  return options.map(
    (option: ISelectOption): ISelectOption => ({
      ...option,
      clearableValue: false
    })
  )
}
