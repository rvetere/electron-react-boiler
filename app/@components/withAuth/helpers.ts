import config from '@config'
import logger from '@helpers/logger'
import { IAuthState } from '@reducers/types'
import jwt from 'jsonwebtoken'
import { ITokenData } from './types'

export const verifyToken = (token: string): Promise<ITokenData> =>
  new Promise(
    (resolve: (decoded?: ITokenData) => void): void => {
      jwt.verify(
        token,
        config.secret,
        (error: Error, decoded: ITokenData): void => {
          if (error) {
            logger.error(error.message)
            return resolve()
          }

          resolve(decoded)
        }
      )
    }
  )

export const isAuthorized = (auth: IAuthState): boolean => {
  if (auth && auth.token) {
    verifyToken(auth.token).then(
      (_tokenData: ITokenData): void => {
        // TODO: dispatch signOut if invalid!
      }
    )

    return true
  }

  return false
}
