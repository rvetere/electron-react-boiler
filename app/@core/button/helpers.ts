import classNames from 'class-names'
import React, { ReactNode } from 'react'
import styles from './button.css'
import { IButtonProps, IButtonStylings } from './types'

export const getIconPadding = (children: ReactNode): boolean => {
  if (!children) {
    return false
  }

  const count = React.Children.count(children)
  let iconPadding = React.isValidElement(children[0]) && count > 1

  if (count > 1 && children[1].props) {
    iconPadding = children[1].props.className !== 'ssrOnly'
  }

  return iconPadding
}

export const getButtonStylings = (children: ReactNode, props: IButtonProps): IButtonStylings => {
  const { primary, secondary, outline, small, big, inverted, block, mobileNoBlock, loading, fullSize, ...otherProps }: IButtonProps = props
  const className = classNames(styles.button, {
    [styles.disabled]: props.disabled,
    [styles.primary]: primary,
    [styles.secondary]: secondary,
    [styles.small]: small,
    [styles.outline]: outline,
    [styles.big]: big,
    [styles.inverted]: inverted,
    [styles.iconPadding]: getIconPadding(children),
    [styles.block]: block,
    [styles.mobileNoBlock]: mobileNoBlock,
    [styles.loading]: loading,
    [styles.fullSizeOverlay]: fullSize
  })
  return { className, otherProps }
}
