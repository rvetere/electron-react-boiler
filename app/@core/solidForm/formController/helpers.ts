import {
  CurryUpdateState,
  IFormControllerField,
  IFormControllerFields,
  IFormControllerPayload,
  IFormControllerState,
  IFormValues
} from '../formController/types'

export const isFormValid = (fields: IFormControllerFields): boolean => Object.values(fields).every(({ isValid }: IFormControllerField): boolean => isValid)

export const getFormValues = (fields: IFormControllerFields): IFormValues =>
  Object.keys(fields).reduce(
    (accumulatedValues: {}, name: string): IFormValues => ({
      ...accumulatedValues,
      [name]: fields[name].value
    }),
    {}
  )

export const updateFormFields = ({ name, ...submittedFields }: IFormControllerPayload): CurryUpdateState => ({
  fields: previousFields
}: Pick<IFormControllerState, 'fields'>): Pick<IFormControllerState, 'fields'> => ({
  fields: {
    ...previousFields,
    [name]: {
      ...submittedFields
    }
  }
})
