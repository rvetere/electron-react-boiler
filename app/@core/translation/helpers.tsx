import { Fragment, ReactNode } from 'react'

export class TranslationError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'TranslationError'
    this.message = message
  }
}

export const containsPlaceholder = (value: string): boolean => value.indexOf('[') !== -1
export const getPlaceholderIndex = (value: string): number => +value.slice(value.indexOf('[') + 1, value.indexOf(']'))
export const getTranslationParts = (translation: string): string[] =>
  translation
    .replace(/{/g, '[')
    .replace(/}/g, ']')
    .split(/(\[\d+\])/g)

export const getInterpolatedTranslation = (translation: string, interpolations: ReactNode[]): ReactNode => {
  const parts = getTranslationParts(translation)

  return parts.map((val: string, i: number) => {
    if (containsPlaceholder(val)) {
      const interpolation = interpolations[getPlaceholderIndex(val)]

      if (typeof interpolation === 'string' || typeof interpolation === 'number') {
        return interpolation
      }

      return <Fragment key={i}>{interpolation}</Fragment>
    }

    return val
  })
}
