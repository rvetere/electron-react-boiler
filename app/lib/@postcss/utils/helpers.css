.textLeft {
  text-align: left;
}

.textRight {
  text-align: right;
}

.pullLeft {
  float: left;
}

.pullRight {
  float: right;
}

.clearfix {
  &::before,
  &::after {
    content: ' ';
    display: table;
  }

  &::after {
    clear: both;
  }
}

.errorMessage {
  position: relative;
  z-index: $zIndexTeaser;
  margin-top: $spacingBase;
  margin-bottom: $spacingBase;
  color: $errorColor;
}

.noPadding {
  padding: 0 !important; /* stylelint-disable-line */
}

.delimiterAfter {
  &::after {
    content: '';
    position: absolute;
    bottom: 0;
    width: 100%;
    height: 1px;
    background: $lightDelimiterColor;
  }
}

.noDelimiterAfterLast:last-child {
  &::after {
    display: none;
  }
}

.delimiterBefore {
  &::before {
    content: '';
    position: absolute;
    top: 0;
    width: 100%;
    height: 1px;
    background: $lightDelimiterColor;
  }
}

/* ASPECT RATIOS */
.aspectRatio {
  position: relative;

  &::before {
    content: ' ';
    display: block;
  }
}

.aspectRatioContent {
  position: absolute;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
}

.aspectRatio2x1 {
  composes: aspectRatio;

  &::before {
    padding-top: 50%;
  }
}

.aspectRatio3x2 {
  composes: aspectRatio;

  &::before {
    padding-top: 66.6666666666%;
  }
}

.aspectRatio4x3 {
  composes: aspectRatio;

  &::before {
    padding-top: 75%;
  }
}

.aspectRatio16x9 {
  composes: aspectRatio;

  &::before {
    padding-top: 56.25%;
  }
}

/* stylelint-disable */
.loadingSpinner,
:global(.Select-loading) {
  /* stylelint-enable */
  display: inline-block;
  box-sizing: border-box;
  vertical-align: middle;
  width: 18px;
  height: 18px;
  border: solid 1px transparent;
  border-top-color: $primaryColor;
  border-left-color: $primaryColor;
  border-radius: 50%;
  animation: spinner 500ms linear infinite;
}

@keyframes spinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}

/* TODO: try to remove !important after refactoring of button component! check in footer */
.hiddenXs {
  @media (max-width: $breakpointXsMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenSm {
  @media (min-width: $breakpointSmMin) and (max-width: $breakpointSmMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenMd {
  @media (min-width: $breakpointMdMin) and (max-width: $breakpointMdMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenLg {
  @media (min-width: $breakpointLgMin) and (max-width: $breakpointLgMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenXl {
  @media (min-width: $breakpointXlMin) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenXsUp {
  @media (min-width: $breakpointSmMin) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenSmDown {
  @media (max-width: $breakpointSmMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenSmUp {
  @media (min-width: $breakpointSmMin) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenMdDown {
  @media (max-width: $breakpointMdMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenMdUp {
  @media (min-width: $breakpointMdMin) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenLgDown {
  @media (max-width: $breakpointMdMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenLgUp {
  @media (min-width: $breakpointLgMin) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenXlDown {
  @media (max-width: $breakpointLgMax) {
    display: none !important; /* stylelint-disable-line */
  }
}

.hiddenXlUp {
  @media (min-width: $breakpointXlMin) {
    display: none !important; /* stylelint-disable-line */
  }
}
