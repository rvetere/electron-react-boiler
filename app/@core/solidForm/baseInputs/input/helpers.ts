import { InputValue, ISelectOption } from '@core/solidForm'
import { ValidatorFn } from '@core/solidForm/validators/types'
import { ICheckboxProps } from '../checkbox/types'
import { IRadioGroupProps } from '../radioGroup/types'
import { ISelectProps } from '../select/types'
import { ITextAreaProps } from '../textArea/types'
import { IInputTextProps, IInputValidationPayload, IInputValidationResult } from './types'

export const applyValidators = (value: InputValue | ISelectOption[], validators: ValidatorFn[]): IInputValidationResult =>
  validators.reduce(
    (result: IInputValidationResult, validator: ValidatorFn): IInputValidationResult => {
      const { isValid, message } = validator(value)

      if (result.isValid !== false) {
        result.isValid = isValid
      }

      if (message && !isValid) {
        result.messages = [...result.messages, message]
      }

      return result
    },
    { isValid: true, messages: [] }
  )

export const getNativeValidation = ({ current }: React.RefObject<HTMLInputElement | HTMLTextAreaElement | HTMLSelectElement>): IInputValidationResult => {
  return {
    isValid: current ? current.validity.valid : true,
    messages:
      current && !current.validity.valid
        ? [
            {
              message: current.validationMessage
            }
          ]
        : []
  }
}

export const validateInput = ({ value, validators, inputRef }: IInputValidationPayload): IInputValidationResult => {
  const { messages, isValid } = validators ? applyValidators(value, validators) : { isValid: true, messages: [] }
  const { messages: nativeMessages, isValid: nativeIsValid } = inputRef ? getNativeValidation(inputRef) : { isValid: true, messages: [] }

  return {
    isValid: isValid && nativeIsValid,
    messages: [...messages, ...nativeMessages]
  }
}

export const hasValidation = (props: ICheckboxProps | IInputTextProps | IRadioGroupProps | ISelectProps | ITextAreaProps): boolean => {
  // @ts-ignore
  return props.required || (props.validators && props.validators.length > 0)
}
