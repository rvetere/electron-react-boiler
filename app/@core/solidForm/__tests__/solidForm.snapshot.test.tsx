/* eslint-env jest */
import { selectOptions } from '@misc/styleguide/solidForm/sample/helpers'
import { getByClass } from '@testing/classQueries'
import { ReactNode } from 'react'
import { cleanup, fireEvent, render, RenderResult, waitForElement } from 'react-testing-library'
import { Checkbox, FormController, IFormControllerActions, InputText, IRadioProps, RadioGroup, RadioOption, Select, TextArea } from '..'

enum InputTypes {
  InputText = 'InputText',
  Select = 'Select',
  Checkbox = 'Checkbox',
  RadioGroup = 'RadioGroup',
  TextArea = 'TextArea'
}

type FormRenderResult = RenderResult & {
  onUpdateMock: jest.Mock<{}>
  onResetMock: jest.Mock<{}>
  renderProps: IFormControllerActions
}

describe('FormController', () => {
  afterEach(cleanup)

  test('Integration: InputText child of FormController must display validationMessages on submit attempt', async () => {
    const { container, onUpdateMock, onResetMock, renderProps } = renderAndSubmitForm(InputTypes.InputText)
    let validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))
    expect(validationDisplay).toBeTruthy()

    // check update functionality
    const inputEl = container.querySelector('input')
    fireEvent.change(inputEl!, { target: { value: 'a change' } })

    expect(onUpdateMock).toBeCalledTimes(1)

    // check reset functionality
    renderProps.resetForm()
    expect(onResetMock).toBeCalledTimes(1)
    expect(inputEl!.value).toEqual('')

    // check validationDisplay logic again after reset
    const formEl = container.querySelector('form')
    fireEvent.submit(formEl!)

    validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))
    expect(validationDisplay).toBeTruthy()
  })

  test('Integration: Checkbox child of FormController must display validationMessages on submit attempt', async () => {
    const { container } = renderAndSubmitForm(InputTypes.Checkbox)
    const validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))
    expect(validationDisplay).toBeTruthy()
  })

  test('Integration: RadioGroup child of FormController must display validationMessages on submit attempt', async () => {
    const { container } = renderAndSubmitForm(InputTypes.RadioGroup)
    const validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))
    expect(validationDisplay).toBeTruthy()
  })

  test('Integration: Select child of FormController must display validationMessages on submit attempt', async () => {
    const { container } = renderAndSubmitForm(InputTypes.Select)
    const validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))
    expect(validationDisplay).toBeTruthy()
  })

  test('Integration: TextArea child of FormController must display validationMessages on submit attempt', async () => {
    const { container } = renderAndSubmitForm(InputTypes.TextArea)
    const validationDisplay = await waitForElement(() => getByClass(container, 'validationMessage'))
    expect(validationDisplay).toBeTruthy()
  })
})

const getInputByType = (type: InputTypes, renderProps: IFormControllerActions): ReactNode => {
  switch (type) {
    case InputTypes.InputText:
      return <InputText name="inputText" label="InputText" inputProps={renderProps.inputProps} required />

    case InputTypes.TextArea:
      return <TextArea name="textArea" label="TextArea" inputProps={renderProps.inputProps} required />

    case InputTypes.Checkbox:
      return (
        <Checkbox name="checkbox" inputProps={renderProps.inputProps} required>
          a checkbox
        </Checkbox>
      )

    case InputTypes.RadioGroup:
      return (
        <RadioGroup name="radioGroup" inputProps={renderProps.inputProps} required>
          {(radioProps: IRadioProps): ReactNode => (
            <>
              <RadioOption disabled value="test1" radioProps={radioProps}>
                test 1 Label
              </RadioOption>
              <RadioOption value="test2" radioProps={radioProps}>
                test 2 Label
              </RadioOption>
            </>
          )}
        </RadioGroup>
      )

    case InputTypes.Select:
      return <Select name="select" label="Select" inputProps={renderProps.inputProps} required options={selectOptions} />
  }
}

const renderAndSubmitForm = (type: InputTypes = InputTypes.InputText): FormRenderResult => {
  const onSubmitMock = jest.fn()
  const onUpdateMock = jest.fn()
  const onInvalidMock = jest.fn()
  const onResetMock = jest.fn()

  let resultRenderProps: IFormControllerActions | null = null
  const result = render(
    <FormController onSubmit={onSubmitMock} onUpdate={onUpdateMock} onInvalid={onInvalidMock} onReset={onResetMock}>
      {(renderProps: IFormControllerActions): ReactNode => {
        expect(renderProps).toBeDefined()
        resultRenderProps = renderProps
        return getInputByType(type, renderProps)
      }}
    </FormController>
  )

  const formEl = result.container.querySelector('form')
  fireEvent.submit(formEl!)

  expect(onInvalidMock).toBeCalledTimes(1)
  expect(result.container).toMatchSnapshot()

  return { ...result, onUpdateMock, onResetMock, renderProps: resultRenderProps! }
}
