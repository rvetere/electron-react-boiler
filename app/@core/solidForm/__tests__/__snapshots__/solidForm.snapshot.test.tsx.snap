// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormController Integration: Checkbox child of FormController must display validationMessages on submit attempt 1`] = `
<div>
  <form>
    <label
      class="formField inputElement checkbox error"
    >
      <input
        aria-describedby="checkboxValidationMessage"
        class="htmlInput"
        name="checkbox"
        required=""
        type="checkbox"
      />
      <span
        class="checkboxSvg inputSvg"
      >
        <test-file-stub
          classname="digicon"
          height="18"
          width="18"
        />
      </span>
      a checkbox
    </label>
    <div
      class="wrapper"
      id="checkboxValidationMessage"
    >
      <p
        class="validationMessage"
      >
        Constraints not satisfied
      </p>
    </div>
  </form>
</div>
`;

exports[`FormController Integration: InputText child of FormController must display validationMessages on submit attempt 1`] = `
<div>
  <form>
    <div
      class="formField wrapper error"
    >
      <label
        class="formLabel label"
      >
        <span>
          InputText
        </span>
        <input
          aria-describedby="inputTextValidationMessage"
          name="inputText"
          required=""
          value=""
        />
        <div
          class="wrapper"
          id="inputTextValidationMessage"
        >
          <p
            class="validationMessage"
          >
            Constraints not satisfied
          </p>
        </div>
      </label>
    </div>
  </form>
</div>
`;

exports[`FormController Integration: RadioGroup child of FormController must display validationMessages on submit attempt 1`] = `
<div>
  <form>
    <div
      class="formField"
    >
      <label
        class="radioOption radio inputElement disabled error"
      >
        <input
          aria-describedby="radioGroupValidationMessage"
          class="htmlInput"
          disabled=""
          name="radioGroup"
          type="radio"
          value="test1"
        />
        <span
          class="radioSvg inputSvg"
        >
          <test-file-stub
            classname="digicon"
            height="18"
            width="18"
          />
        </span>
        test 1 Label
      </label>
      <label
        class="radioOption radio inputElement error"
      >
        <input
          aria-describedby="radioGroupValidationMessage"
          class="htmlInput"
          name="radioGroup"
          type="radio"
          value="test2"
        />
        <span
          class="radioSvg inputSvg"
        >
          <test-file-stub
            classname="digicon"
            height="18"
            width="18"
          />
        </span>
        test 2 Label
      </label>
      <div
        class="wrapper"
        id="radioGroupValidationMessage"
      >
        <p
          class="validationMessage"
        >
          Required
        </p>
      </div>
    </div>
  </form>
</div>
`;

exports[`FormController Integration: Select child of FormController must display validationMessages on submit attempt 1`] = `
<div>
  <form>
    <div
      class="formField"
    >
      <label
        class="formLabel label"
      >
        <span>
          Select
        </span>
        <div
          class="wrapper error"
        >
          <select
            aria-describedby="selectValidationMessage"
            class="select"
            id="select"
            name="select"
            required=""
          >
            <option
              value="0"
            >
              Option 1
            </option>
            <option
              value="1"
            >
              Option 2
            </option>
            <option
              value="2"
            >
              Option 3
            </option>
            <option
              value="3"
            >
              Option 4
            </option>
            <option
              value="4"
            >
              Option 5
            </option>
            <option
              value="5"
            >
              Option 6
            </option>
            <option
              value="6"
            >
              Option 7
            </option>
            <option
              value="7"
            >
              Option 8
            </option>
            <option
              value="8"
            >
              Option 9
            </option>
            <option
              value="9"
            >
              Option 10
            </option>
            <option
              value="10"
            >
              Option 11
            </option>
            <option
              value="11"
            >
              Option 12
            </option>
            <option
              value="12"
            >
              Option 13
            </option>
            <option
              value="13"
            >
              Option 14
            </option>
            <option
              value="14"
            >
              Option 15
            </option>
            <option
              value="15"
            >
              Option 16
            </option>
            <option
              value="16"
            >
              Option 17
            </option>
            <option
              value="17"
            >
              Option 18
            </option>
            <option
              value="18"
            >
              Option 19
            </option>
            <option
              value="19"
            >
              Option 20
            </option>
          </select>
          <test-file-stub
            classname="digicon"
            height="16"
            width="16"
          />
        </div>
        <div
          class="wrapper"
          id="selectValidationMessage"
        >
          <p
            class="validationMessage"
          >
            Constraints not satisfied
          </p>
        </div>
      </label>
    </div>
  </form>
</div>
`;

exports[`FormController Integration: TextArea child of FormController must display validationMessages on submit attempt 1`] = `
<div>
  <form>
    <div
      class="formField wrapper error"
    >
      <label
        class="formLabel label"
      >
        <span>
          TextArea
        </span>
        <textarea
          aria-describedby="textAreaValidationMessage"
          name="textArea"
          required=""
        />
        <div
          class="wrapper"
          id="textAreaValidationMessage"
        >
          <p
            class="validationMessage"
          >
            Constraints not satisfied
          </p>
        </div>
      </label>
    </div>
  </form>
</div>
`;
